poller.py

```diff
@@ -73,50 +59,68 @@ def _process_input(self, input_q: dict) -> None:
             chat_ids.append(chat_id)
             log_msg = f'{log_msg}, {str(chat_id)}' if log_msg else f'Processing messages for chats: {str(chat_id)}'
 
+        self.chat_events = {chat_id: [asyncio.Event() for _ in range(len(chat))] for chat_id, chat in buffer.items()}
+        for events in self.chat_events.values():
+            events[-1].set()
+
         if log_msg:
-            log.info(log_msg)
+            self.log.info(log_msg)
 
         # "slices" of replicas from all conversations, each slice contains replicas from different conversation
         batched_chats = zip_longest(*chats, fillvalue=None)
 
-        for chats_batch in batched_chats:
-            utts_batch = [(chat_ids[utt_id], utt) for utt_id, utt in enumerate(chats_batch) if utt]
-            j = self.config['infer_batch_length']
-            chunked_utts_batch = [utts_batch[i * j:(i + 1) * j] for i in range((len(utts_batch) + j - 1) // j)]
-
-            for chunk in chunked_utts_batch:
-                batch = list(zip(*chunk))
-                ids_batch = list(batch[0])
-                utts_batch = list(batch[1])
-                response_batch = self.agent(utts_batch, ids_batch)
-
-                for resp in zip(ids_batch, response_batch):
-                    self._send_results(*resp)
+        tasks = (self.loop.create_task(self._process_chats_batch(chats_batch,
+                                                                 msg_id,
+                                                                 chat_ids)) for msg_id, chats_batch in enumerate(batched_chats))
+        await asyncio.gather(*tasks)
+
+    async def _process_chats_batch(self, chats_batch: list, msg_id: int, chat_ids: list) -> None:
```

Comment:
Here, before and further use type annotation from `typing` if possible: for example `List[int]` or `Dict[str, List[str]]`.

Reply:
Done for lists. Dicts left unchanged due to heterogeneous content.